### ISSUE TITLE
Time zone parsing error

### ISSUE THREAD

**Issue by andjd** [eventable/vobject#18]
I'm trying to generate an iCalendar string with timezone data on the dtstart element.  The code runs fine when I pass in a naive python datetime object, but when I pass in datetime object with a pytz/tzinfo component, the serialize method crashes with the following error: 

```
.../python2.7/site-packages/pytz/tzinfo.pyc in localize(self, dt, is_dst)
    325             # If we refuse to guess, raise an exception.
    326             if is_dst is None:
--> 327                 raise NonExistentTimeError(dt)
    328 
    329             # If we are forcing the pre-DST side of the DST transition, we

NonExistentTimeError: 2000-04-02 02:00:00
```

following the stack trace, the error occurs after vobject calls pytz during the serialize call:

```
.../python2.7/site-packages/vobject/icalendar.pyc in test(dt)
   1863     assert transitionTo in ('daylight', 'standard')
   1864     if transitionTo == 'daylight':
-> 1865         def test(dt): return tzinfo.dst(dt) != zeroDelta
   1866     elif transitionTo == 'standard':
   1867         def test(dt): return tzinfo.dst(dt) == zeroDelta
```

This smells like a bug in the internal implementation to me, especially since the datetime returned in the error is over 15 years removed from the dates I'm working with.

**Comment by damangio** [comment]
I don't know if this is incorrect behavior or not, but here is a workaround: http://stackoverflow.com/questions/21128695/django-ical-with-vobject-issue-with-pytz

**Comment by untitaker** [comment]
I'm experiencing the same issue with [this ICS](https://gist.github.com/untitaker/95155a496ee9d29cb067de1f57a3feb4). I noticed that the error doesn't occur if pytz isn't installed.

Also, here's a trivial testcase that uses the file above:

```diff
diff --git a/tests.py b/tests.py
index 2073930..8be3d83 100644
--- a/tests.py
+++ b/tests.py
@@ -440,6 +440,11 @@ class TestGeneralFileParsing(unittest.TestCase):
              ['NEXT', 'Nope'], ['BAR']]
         )
 
+    def test_foo(self):
+        cal = get_test_file("foo.ics")
+        cal = base.readOne(cal)
+        cal.serialize()
+
 
 class TestVcards(unittest.TestCase):
     """
```

**Comment by medmunds** [comment]
@untitaker I *think* [your ICS](https://gist.github.com/untitaker/95155a496ee9d29cb067de1f57a3feb4) is technically invalid, if I'm reading [RFC 5545](https://tools.ietf.org/html/rfc5545#section-3.2.19) correctly (check out sections 3.2.19 and 3.3.5 "Form # 3"). The ICS refers to `TZID=Australia/Sydney`, but is missing a `VTIMEZONE` block to define that timezone. (The RFC doesn't define any standard timezones, other than UTC, and seems to go out of its way to avoid mentioning the IANA/Olson tzdb by name. So you have to include the rules for any timezone used by your events.) Do you know what generated that ICS?

You're running into some undocumented vobject behavior that tries to use pytz to interpret missing VTIMEZONEs [(this code)](https://github.com/eventable/vobject/blob/0.9.2/vobject/icalendar.py#L55-L64). That behavior is also, arguably, technically incorrect. But it's in there--and probably helpful in some cases. The problem is, once the pytz timezone is attached, you run into this bug (or #32) on serialization, when vobject helpfully tries to generate the missing VTIMEZONE block for you.

(BTW, I think you could make your ICS technically correct by changing it to `TZID=/Australia/Sydney` -- the prefix `/` makes the TZID refer to a ["globally defined time zone registry (when such registry is defined)."](https://tools.ietf.org/html/rfc5545#section-3.2.19) And vobject should really only use pytz to resolve TZIDs that start with `/`. But I haven't run into any calendar apps that actually generate or respect this part of the RFC on import, so this is just a technical argument without practical merit.)

**Comment by untitaker** [comment]
@medmunds You're right. I'm not sure about the origin of that ICS. It was just a testcase anyway and might have been mangled by me a long time ago when I didn't know what I was doing.

@geier might know if there are any producers that actually reference TZIDs without defining them. I suspect there are a few.

**Comment by geier** [comment]
I have seen some, but not a lot of them (but IIRC all with Olson DB identifiers).

**Comment by wpercy** [comment]
I believe this was resolved by #33
